src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,147 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        tail = new Node<>(tail, value, null);
+        if (head == null) {
+            head = tail;
+        } else {
+            tail.prev.next = tail;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) {
+            add(value);
+            return;
+        }
+        checkIndex(index);
+        if (index == 0) {
+            head = new Node<>(null, value, head);
+            head.next.prev = head;
+            size++;
+            return;
+        }
+        Node<T> nextNode = findNodeByIndex(index);
+        Node<T> newNode = new Node<>(nextNode.prev, value, nextNode);
+        newNode.prev.next = newNode;
+        nextNode.prev = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T currentElement : list) {
+            add(currentElement);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return findNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> currentNode = findNodeByIndex(index);
+        T oldValue = currentNode.value;
+        currentNode.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> nodeToRemove = findNodeByIndex(index);
+        unlink(nodeToRemove);
+        return nodeToRemove.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = head;
+        while (currentNode != null) {
+            if (currentNode.value == object
+                    || currentNode.value != null
+                    && currentNode.value.equals(object)) {
+                unlink(currentNode);
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index > size - 1) {
+            throw new IndexOutOfBoundsException("IndexOutOfBoundsException on index " + index);
+        }
+    }
+
+    private Node<T> findNodeByIndex(int index) {
+        checkIndex(index);
```

Comment:
we call method `checkIndex(index)` here, so we can remove `checkIndex(index)` from all other methods (get, set,...)